### PR TITLE
Add hypermedia API to replace XML-RPC and simple

### PR DESCRIPTION
**Pull request by asmacdo**
The working plan and reasoning is documented on this etherpad, comments and additions are encouraged:  https://pad.sfconservancy.org/p/hypermedia_api_design

This work is based on the discussion from:
https://github.com/pypa/warehouse/issues/284

Introduces new dependencies:

- apispec==0.37.0 : Used to generate an api spec at /api/
- marshmallow==3.0.0b10 : Used to serialize responses
- PyYAML==3.12 : Dependency of apispec
- 

All new endpoints are added to a new subdomain, "api".

### CODE REVIEW

**Comment by ewdurbin** [comment]
@asmacdo Awesome to see the first iteration of a next generation API coming together. Thanks for your contributions during sprints.

We can still likely roll this out under the new subdomain, but do require 100% test coverage for merges.

File: requirements/main.in [hunk 1]
```diff
@@ -36,6 +38,7 @@ pyramid_retry>=0.3
 pyramid_rpc>=0.7
 pyramid_services
 pyramid_tm>=0.12
+PyYaml
```
Comment (merwok) [inline requirements/main.in]: If this is a dependency of apispec, it shouldn’t be in the `in` file.

File: warehouse/api/schema.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    filename = fields.Str()
+    packagetype = fields.Str()
+    python_version = fields.Str()
+    has_sig = fields.Bool(attribute="has_signature")
+    comment_text = fields.Str()
+    md5_digest = fields.Str()
+    digests = fields.Method("get_digests")
+    size = fields.Int()
+    upload_time = fields.Function(
+        lambda obj: obj.upload_time.strftime("%Y-%m-%dT%H:%M:%S")
```
Comment (merwok) [inline warehouse/api/schema.py]: No timezone?

File: warehouse/api/schema.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+import datetime
+
+from marshmallow import Schema, fields
+
+
+class File(Schema):
+    filename = fields.Str()
+    packagetype = fields.Str()
+    python_version = fields.Str()
+    has_sig = fields.Bool(attribute="has_signature")
```
Comment (merwok) [inline warehouse/api/schema.py]: Why change the name?

File: warehouse/api/schema.py (hunk 2 above)
Comment (sloria) [reply warehouse/api/schema.py]: Drive-by comment: You can use `fields.DateTime`. You can pass an optional format:

```python
import datetime as dt

import pytz
from marshmallow import Schema, fields

class File(Schema):
    upload_time = fields.DateTime(format="%Y-%m-%dT%H:%M:%S")


schema = File()
timezone_aware_dt = dt.datetime.utcnow()
timezone_aware_dt.replace(tzinfo=pytz.UTC)
print(schema.dump({'upload_time': timezone_aware_dt}))
# {'upload_time': '2018-06-01T13:29:25'}
```

If you don't pass a format, the default behavior is to format as ISO8601.

```python
import datetime as dt

import pytz
from marshmallow import Schema, fields

class File(Schema):
    upload_time = fields.DateTime()


schema = File()
timezone_aware_dt = dt.datetime.utcnow()
timezone_aware_dt.replace(tzinfo=pytz.UTC)
print(schema.dump({'upload_time': timezone_aware_dt}))
# {'upload_time': '2018-06-01T13:31:08.732522+00:00'}
```

File: warehouse/api/schema.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+            "summary",
+            "version",
+            "files_url",
+        )
+        ordered = True
+
+
+class Project(Schema):
+    url = fields.Method("get_detail_url")
+    releases_url = fields.Method("get_releases_url")
```
Comment (sloria) [inline warehouse/api/schema.py]: Drive-by comment: Since serializing URLs is a common usage, you might want to create a custom field. It could look something like

```python
class Project(Schema):
    releases_url = RouteURL("api.views.projects.releases",
                            name="<normalized_name>")
```

For an example of how this might be implemented, check out flask-marshmallow's equivalent: https://github.com/marshmallow-code/flask-marshmallow/blob/7e594dc89ba853946ef1593e9ac19850a5ec9d2a/flask_marshmallow/fields.py#L39

File: warehouse/api/schema.py (hunk 2 above)
Comment (asmacdo) [reply warehouse/api/schema.py]: @merwok, we left out the timezone because that's what the JSON api does. However, the XML-RPC api does include a tz. I'll add tz to this field on the next push.

File: warehouse/api/schema.py (hunk 3 above)
Comment (asmacdo) [reply warehouse/api/schema.py]: `has_sig` is what the JSON API uses, so I started there. I prefer to keep the names consistent with the db where possible, I'll change it to has_signature.

**Comment by brainwane** [comment]
@asmacdo Thank you for working on this!

@dstufft @di have you had a chance to take a look at this branch, or the Etherpad?

**Comment by di** [comment]
I haven't yet, sorry, but it's on my list.

**Comment by brainwane** [comment]
@di Do you think you'll have time to review this during August?

**Comment by asmacdo** [comment]
This code has rotted a bit, so I'm closing.
